### ISSUE TITLE
DeerConnect:  Improve spam mitigation techniques on contact form

### ISSUE THREAD

**Issue by lupinia** [lupinia/awi#47]
Ideally implemented with regular expression checking for certain keywords, but blacklisting certain IP addresses would also be handy.

**Comment by lupinia** [comment]
Clarifying some thoughts on this:  

- I definitely do *not* want to implement ReCAPTCHA.  Google has effectively monopolized spam mitigation, to the point that ReCAPTCHA shows up even in places where it makes no sense, and while it's undeniably effective, I refuse on principle.
- Trying to use a CAPTCHA is, in general, not a great solution.  This will have to rely on backend processing.
- The difficulty with a keyword list is maintaining it, so I'll need a simple way to add new spam flags as new attacks emerge.  
- Blocking IP addresses is a challenge, because I don't want to rely on a database lookup for that; I'm currently doing it in the nginx config, but that's difficult to maintain by hand, and I don't want to grant the web application permission to modify its own nginx config files.  Maybe I can automate the workflow somehow, or leverage something like fail2ban (which could also potentially be useful for the login form).  
- In the event of a false positive, I'll need to provide an alternate means of reporting the issue.  I'll probably end up directing people to open a ticket in this Github repo.  
- Whatever I do to implement this will require an update to my privacy policy, so since that's kind of a given, I may look into whether there are AWS products that could help without using every message to train some generative AI garbage.

**Comment by lupinia** [comment]
After doing some research, I'm starting to get a clearer idea of specific implementation goals:  

- Add an additional honeypot-like field, visible in the form, just a checkbox labeled something to the effect of "This message is a commercial solicitation".  The text at the top of the field explicitly says that the form may not be used for such communication, and it will be considered spam, but my hypothesis is that the wording of that text, and the relationship to a checkbox on the form, would be too sophisticated for a bot, so they would check the box, but a human who reads the text would not.  Need to decide how long this should lock the offending user out of the form - maybe the first offense locks them out for 24 hours, and the second locks them out permanently.  
- Move the current session-based resubmission timeout check to a cache-based check, using the combination of IP and browser headers.  This would prevent malicious users from simply clearing their cookies or visiting from a different domain to re-submit early.  
- Add a keyword detection system.  Failure by this method should lock the user out of the form permanently.  
- I still need to find a way to block IP addresses at the server level without granting the application code permission to edit its own nginx config files.  On the other hand, IP blocking at that level is something I should probably only do by hand, in the form of IP ranges, so maybe I can just have a "restricted IPs" table that blocks access at the application level.  That's probably fine.

**Comment by lupinia** [comment]
Spam filtering is now implemented!  
- Messages are now checked for keyword matches against a list of known spam links/words/phrases.
- Sender email address is recorded in the event of a spam keyword match, and future mail from the same sender is blocked.
- Sender IP address is permanently blocked for spam keyword matches.  

I need to improve the IP block code to make it apply to the whole site, but that'll be handled in #136.  I've tested this as much as I can - legitimate mail gets through, spam I've already received gets blocked - and everything seems to be working correctly.
